### PR TITLE
Backend support for DO-Sync

### PR DESCRIPTION
**Pull request by jgoizueta**
Story details: https://app.clubhouse.io/cartoteam/story/89202

### CODE REVIEW

**Comment by jgoizueta** [comment]
Pending:
* Unit tests for DoAPIClient & BqClient (I don't see to much value in them, I prefer to start staging tests)
* Account limits (if the estimated size is larger than the user available quota, a subscriptions should be unsyncable)
* DO sync limits are now hardcoded in the code
* Should raise exceptions in cases like creating a sync for an unsyncable subscription? (related: naming of methods raising exceptions)

File: app/services/carto/do_sync_service.rb [hunk 1]
```diff
@@ -26,12 +31,62 @@ def initialize(user)
...
+      subscription = @user.do_subscription(subscription_id)
+
+      if subscription.blank?
+        return {
+          DO_SYNC_STATUS => DO_SYNC_STATUS_UNSYNCABLE,
+          DO_SYNC_UNSYNCABLE_REASON => "Invalid subscription #{subscription_id}"
+        }.with_indifferent_access
+      end
+
+      if subscription['expires_at'] <= Time.now
```
Comment (rafatower) [inline app/services/carto/do_sync_service.rb:43]: to be consistent with the comment and code above: `Time.now < subscription['expires_at']`

File: lib/carto/gcloud_user_settings.rb [hunk 2]
```diff
@@ -1,34 +1,36 @@
...
-    def initialize(user, attributes)
+    def initialize(user)
       @username = user.username
-      @attributes = attributes
     end
 
-    def update
-      if @attributes.present?
-        store
+    def update(attributes)
```
Comment (rafatower) [inline lib/carto/gcloud_user_settings.rb:14]: Slightly off-topic: I still need to test this end-to-end:

I think this causes trouble:
```
# in central
user.server.update_user(gcloud_settings: {})
```
in cartodb: https://rollbar.com/carto/CartoDB/items/41673/

File: spec/services/carto/do_sync_service_spec.rb [hunk 3]
```diff
@@ -32,6 +34,17 @@
     @redis_key = "do:#{@user.username}:datasets"
     $users_metadata.hset(@redis_key, 'bq', bq_datasets.to_json)
 
+    gcloud_settings = {
+      service_account: 'the-service-account',
+      bq_public_project: 'bq-public-project',
+      gcp_execution_project: 'bq-run-project',
+      bq_project: 'bq-project',
+      bq_dataset: 'bq-dataset',
```
Comment (rafatower) [inline spec/services/carto/do_sync_service_spec.rb:42]: minor: invalid dataset name :p

File: app/services/carto/bq_client.rb [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+      client_options.read_timeout_sec = MAX_TIMEOUT
+      request_options = Google::Apis::RequestOptions.new
+      request_options.retries = MAX_CLIENT_RETRIES
+      @service = Google::Apis::BigqueryV2::BigqueryService.new
+      @service.client_options = client_options
+      @service.request_options = request_options
+      @service.authorization = authorizer
+    end
+
+    def query(billing_project_id, sql, dry_run=false)
```
[the remark this answers is not in the document]
Comment (jgoizueta) [reply app/services/carto/bq_client.rb:26]: Right, it's intended for the connector (I brought in this code from db-connectors)

File: app/services/carto/bq_client.rb [hunk 5]
```diff
@@ -0,0 +1,46 @@
+require 'signet/oauth_2/client'
+require 'google/apis/bigquery_v2'
+
+module Carto
+  class BqClient
+    SCOPES = %w(https://www.googleapis.com/auth/cloud-platform
+      https://www.googleapis.com/auth/bigquery).freeze
+    MAX_CLIENT_RETRIES = 3
+    MAX_TIMEOUT = 3600
```
[the remark this answers is not in the document]
Comment (jgoizueta) [reply app/services/carto/bq_client.rb:9]: Yes, that was for the imports (reading query results). Again, I moved this code without looking into it.

**Pull request merged**